### DISCUSSION TITLE
README (style wars meta discussion)

### DISCUSSION THREAD

**Discussion by alwinm** [cholla-hydro/cholla#204] [The Style Wars]
Welcome to the Style Wars. Only by experiencing great division can we unify once again... 

![IMG_2234](https://user-images.githubusercontent.com/5511367/199109572-9ceebc3b-7910-4d95-9b01-37fcfa267a32.jpg)

In this README I discuss our current plan which we can iterate. 

In parallel:
1a) discuss default choice sets (e.g. Google vs Mozilla vs...) for clang-format in 1 **Discussion**
1b) discuss default choice sets (e.g. Google vs Mozilla vs...) for clang-tidy in 1 **Discussion**
1c) for each individual choice (e.g. case naming, spaces/tabs) you feel strongly about, start a new **Discussion** to discuss it
2) TODO: someone will have to learn how to use clang-tidy and clang-format
3a) Have a live group meeting where we finalize choices
3b) Push choices
4) Enforce/Iterate for eternity

Voting and discussing:
I currently imagine we will use 4 emotes when voting:
👎 to be used (scarcely) to show strong disapproval (please explain). 
😕 to show disapproval.
👀  to show you have seen something, you're okay with it, and don't feel strongly one way or another. 
👍 to show strong approval. 

To keep voting organized, make each choice a **Comment** on the **Discussion**. The top level **Comments** should all be voting options or a neutral introduction to the topic. Debates should all happen in **Replies**: general replies under the neutral introduction, option-specific pros/cons under their vote Comment. 

Thus the hierarchy is as follows: **Discussions** (for each issue), **Comments** (for each option), **Replies** (for each pro/con)

FAQ: 
(I'll edit this as appropriate)

**Comment by alwinm** [comment]
Image credit for the Style Wars screen: @bcaddy

**Comment by bcaddy** [comment]
Here's some useful links we might need.

- [clang-format configurator](https://zed0.co.uk/clang-format-configurator/)
- [clang-tidy identifier naming](https://clang.llvm.org/extra/clang-tidy/checks/readability/identifier-naming.html) (@helenarichie you might want to check this out)
- [clang-rename](https://clang.llvm.org/extra/clang-rename.html) refactoring tool. Likely to be very helpful with updating names and such to the new naming conventions. VSC has similar functionality

Other static analyzers we might want to consider:

- [Cppcheck](https://cppcheck.sourceforge.io)
- [Include What You Use](https://include-what-you-use.org)

**Comment by helenarichie** [comment]


**Reply by alwinm** [reply]
I think this would be good as its own Discussion thread. Thoughts?

**Reply by bcaddy** [reply]
I wasn't super clear on the Discussion/comment/thread layout either but I agree that naming convention should be it's own Discussion. I also don't want to have a million discussions. Maybe for now we have a Discussion for formatting and another for naming conventions and comments/threads within those can debate different points. That way all the discourse for a given topic is in one Discussion and not spread out across several. 

Sorry if this contradicts our talk yesterday, Discussion, discussion, comment, and thread are hard to keep track of when they all mean similar but slightly different things.

**Reply by helenarichie** [reply]
Ah I see, I guess I didn't quite understand the format either. Alwin, if you make a separate Discussion thread for naming I will repost this comment in there and delete the one in here!

**Reply by bcaddy** [reply]
@helenarichie, you can make it. Like Issues and PRs it doesn't take admin privileges. (idk if that's new, I just checked to see if I could make one)

**Reply by alwinm** [reply]
Sorry that it was confusing. Thanks all. I agree we shouldn't have a million. I imagine each person will probably contribute between 1-3 things they really really care about, but I do think each topic worth voting on should be its own **Discussion**.

**Reply by bcaddy** [reply]
Link to discussion thread for the deleted comment: https://github.com/cholla-hydro/cholla/discussions/206

**Comment by alwinm** [comment]
Hey y'all, 

Wanted to see what ppl think about how we should use the 👎 emoji in voting: strong negative, negative-leaning but acceptable? Our voting system will take some time to develop :)

**Reply by bcaddy** [reply]
IMO

👍 = "I like this"
😕  = "I don't like this"
👀 = "I have seen this and don't care"
👎 = "I veto this"

**Comment by alwinm** [comment]
Writing a comment here that some additional long-term goal will be:

defining some good practices for doxygen
for our zoo of comments

**Comment by bcaddy** [comment]
# Keyword vs. Symbol Logical Operators

Per @alwinm's suggest and the brief discussion in PR #327 I'm add a place here for discussion/citation of us using keyword (`and`, `not`, `or`, etc) logical operators instead of symbol operators (`&&`, `!`,  `||`, etc). The original discussion has been lost to the annals of Slack but I believe our decision was that keyword operators are more readable and so we should use them moving forward.

**Reply by helenarichie** [reply]
I agree that keyword operators are more readable, so I'd prefer to use those!

**Reply by alwinm** [reply]
Thanks for opening the discussion! My preference is not to enforce in either direction. There are still a lot of (!bool) throughout the Cholla code and the Google style guide prefers punctuation logical operators, even though I personally find keyword operators more readable.

**Reply by bcaddy** [reply]
Yeah, I don’t think it’s super critical. More the kind of thing that if someone notices it in a PR to comment on it

**Reply by alwinm** [reply]
I would consider that a form of enforcement and I also think that automatic enforcement is a better approach than PR comments pushing preferences.
